http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -2,131 +2,229 @@ package org.broadinstitute.dsde.workbench.leonardo
+              for {
+                projectNumber <- projectNumberOptIO.flatMap(_.fold(clusterIamSetupExceptionIO)(IO.pure))
+                // Note that the Dataproc service account is used to retrieve the image, and not the user's
+                // pet service account. There is one Dataproc service account per Google project. For more details:
+                // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
+                dataprocServiceAccountEmail = WorkbenchEmail(
+                  s"service-${projectNumber}@dataproc-accounts.iam.gserviceaccount.com"
+                )
+                _ <- updateGroupMembership(googleGroupsConfig.dataprocImageProjectGroupEmail,
+                                           dataprocServiceAccountEmail,
+                                           createCluster)
+              } yield ()
+            }
+          }
+    }
+
+  /**
+   * Add the Dataproc Worker role in the user's project to the cluster service account, if present.
+   * This is needed to be able to spin up Dataproc clusters using a custom service account.
+   * If the Google Compute default service account is being used, this is not necessary.
+   */
   private def updateClusterIamRoles(googleProject: GoogleProject,
                                     serviceAccountInfo: ServiceAccountInfo,
-                                    create: Boolean): IO[Unit] = {
-    val retryIam: (GoogleProject, WorkbenchEmail, Set[String]) => IO[Boolean] = (project, email, roles) =>
-      IO.fromFuture[Boolean](IO(retryExponentially(when409, s"IAM policy change failed for Google project '$project'") {
+                                    createCluster: Boolean): IO[Unit] = {
+    val retryIam: (GoogleProject, WorkbenchEmail, Set[String]) => IO[Unit] = (project, email, roles) =>
+      IO.fromFuture[Unit](IO(retryExponentially(when409, s"IAM policy change failed for Google project '$project'") {
         () =>
-          if (create) {
-            googleIamDAO.addIamRolesForUser(project, email, roles)
+          val iamUpdate = if (createCluster) {
+            googleIamDAO.addIamRoles(project, email, MemberType.ServiceAccount, roles)
           } else {
-            googleIamDAO.removeIamRolesForUser(project, email, roles)
+            googleIamDAO.removeIamRoles(project, email, MemberType.ServiceAccount, roles)
+          }
+          iamUpdate map { _ =>
+            ()
           }
       }))
 
-    // Add the Dataproc Worker role in the user's project to the cluster service account, if present.
-    // This is needed to be able to spin up Dataproc clusters using a custom service account.
-    // If the Google Compute default service account is being used, this is not necessary.
-    val dataprocWorkerIO = serviceAccountInfo.clusterServiceAccount.map { email =>
-      // Note: don't remove the role if there are existing active clusters owned by the same user,
+    serviceAccountInfo.clusterServiceAccount.map { email =>
```

Comment:
same, use `.traverse`

```
    serviceAccountInfo.clusterServiceAccount.traverse { email =>
      // Note: Don't remove the role if there are existing active clusters owned by the same user,
      // because it could potentially break other clusters. We only check this for the 'remove' case,
      // it's ok to re-add the roles.
      IO.fromFuture(IO(dbRef.inTransaction { _.clusterQuery.countActiveByClusterServiceAccount(email) })).flatMap {
        count =>
          if (count > 0 && !createCluster) {
            IO.unit
          } else {
            retryIam(googleProject, email, Set("roles/dataproc.worker"))
          }
      }
    }.void
```

Reply:
+1, you also use `.traverse_` and git rid of the trailing `.void`, right?